packages/amplify_api_rest/android/src/main/kotlin/com/amazonaws/amplify/amplify_api_rest/AmplifyApiRestPlugin.kt

```diff
@@ -0,0 +1,311 @@
+import com.amplifyframework.api.aws.AWSApiPlugin
+import com.amplifyframework.api.rest.RestOperation
+import com.amplifyframework.api.rest.RestOptions
+import com.amplifyframework.api.rest.RestResponse
+import com.amplifyframework.core.Amplify
+import com.amplifyframework.core.Consumer
+
+import io.flutter.embedding.engine.plugins.FlutterPlugin
+import io.flutter.plugin.common.MethodCall
+import io.flutter.plugin.common.MethodChannel
+import io.flutter.plugin.common.MethodChannel.MethodCallHandler
+import io.flutter.plugin.common.MethodChannel.Result
+import io.flutter.plugin.common.PluginRegistry.Registrar
+import kotlin.reflect.KFunction3
+import kotlin.reflect.KFunction4
+
+class AmplifyApiRestPlugin: FlutterPlugin, MethodCallHandler {
+
+  private lateinit var channel : MethodChannel
+  private var operationsMap : HashMap<String, RestOperation> = HashMap()
+
+  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
+    channel = MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "com.amazonaws.amplify/amplify_api_rest")
+    channel.setMethodCallHandler(this)
+    Amplify.addPlugin(AWSApiPlugin())
+  }
+
+  companion object {
+    val LOG = Amplify.Logging.forNamespace("amplify:flutter:api_rest")
+
+    @JvmStatic
+    fun registerWith(registrar: Registrar) {
+      val channel = MethodChannel(registrar.messenger(), "com.amazonaws.amplify/amplify_api_rest")
+      channel.setMethodCallHandler(AmplifyApiRestPlugin())
+      Amplify.addPlugin(AWSApiPlugin())
+      LOG.info("Added AWSApiPlugin")
+    }
+
+  }
+
+  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
+    channel.setMethodCallHandler(null)
+  }
+
+  @VisibleForTesting
+  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
+    
+    if(call.method == "cancel"){
+      onCancel(result, (call.arguments as String))
+      return
+    }
+    
+    var arguments : Map<String, Any>
+
+    try {
+      arguments = checkArguments(call.arguments as Map<String, Any>)
+    }
+    catch(e: java.lang.Exception){
+      result.error("WrongMapInput", "The provided map input was not correct", e.message)
+      return
+    }
+    
+    when (call.method) {
+      "get" -> onGet(result, arguments)
+      "post" -> onPost(result, arguments)
+      "put" -> onPut(result, arguments)
+      "delete" -> onDelete(result, arguments)
+      else -> result.notImplemented()
+    }
+  }
+
+  private fun restFunctionHelper(
+          @NonNull flutterResult: Result,
+          @NonNull request: Map<String, *>,
+          @NonNull function3: KFunction3<
+                  @ParameterName(name = "restOptions") RestOptions,
+                  @ParameterName(name = "restConsumer") Consumer<RestResponse>,
+                  @ParameterName(name = "exceptionConsumer") Consumer<ApiException>,
+                  RestOperation?>,
+          @NonNull function4: KFunction4<
```

Comment:
What are the `KFunction3` and `KFunction4`? If you are modeling some functional behavior in an interface/type, I suggest creating your own with a semantic name, as opposed to reusing one of Kotlin's internal definitions.

Reply:
cool I used a typealias: 

typealias RestAPIFunction3 = KFunction3<
        @ParameterName(name = "restOptions") RestOptions,
        @ParameterName(name = "restConsumer") Consumer<RestResponse>,
        @ParameterName(name = "exceptionConsumer") Consumer<ApiException>,
        RestOperation?>

Reply:
^^ Ahh, yes. I think this will solve your type inference issue, later, where you had `function3(...)` and I suggested omitting the `Consumer {`? The `RestAPIFunction3` has the type baked in, now, so the inference should work properly.